In [1]:
import pandas as pd
import numpy as np 

C:\Users\HP\anaconda3\Lib\site-packages\pandas\core\computation\expressions.py:22: UserWarning: Pandas requires version '2.10.2' or newer of 'numexpr' (version '2.8.7' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
C:\Users\HP\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:56: UserWarning: Pandas requires version '1.4.2' or newer of 'bottleneck' (version '1.3.7' currently installed).
  from pandas.core import (


In [2]:
orders = pd.read_csv(r"D:\Retail 360\Dataset\Raw\orders.csv")

In [3]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [4]:
orders.shape

(99441, 8)

In [5]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [6]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [7]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 21.9 MB


In [8]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [9]:
orders.duplicated().sum()

0

In [10]:
orders.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522


In [11]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [12]:
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

## Orders Table - Exploration Summary

- Total Rows: 99,441
- Total Columns: 8
- Duplicate Rows: 0

### Missing Values
- order_approved_at: 160
- order_delivered_carrier_date: 1,783
- order_delivered_customer_date: 2,965

### Observations
- All date columns are stored as strings.
- Missing values are expected because some orders were cancelled or not delivered.
- No duplicate rows found.

In [14]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]
for col in date_columns:
    orders[col] = pd.to_datetime(orders[col])

In [15]:
orders.dtypes

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

### Cleaning Step 1: Datetime Conversion

Converted the following columns from string to datetime format:

- order_purchase_timestamp
- order_approved_at
- order_delivered_carrier_date
- order_delivered_customer_date
- order_estimated_delivery_date

**Reason:**
Datetime format is required for time-based analysis such as monthly sales, delivery duration, order trends, and business reporting.

In [17]:
orders["delivery_days"] = (
    orders["order_delivered_customer_date"] - 
    orders["order_purchase_timestamp"]
).dt.days

In [18]:
orders[[
    "order_delivered_customer_date",
    "order_purchase_timestamp",
    "delivery_days"
]].head()

,order_delivered_customer_date,order_purchase_timestamp,delivery_days
0,2017-10-10 21:25:13,2017-10-02 10:56:33,8.0
1,2018-08-07 15:27:45,2018-07-24 20:41:37,13.0
2,2018-08-17 18:06:29,2018-08-08 08:38:49,9.0
3,2017-12-02 00:28:42,2017-11-18 19:28:06,13.0
4,2018-02-16 18:17:02,2018-02-13 21:18:39,2.0


In [19]:
orders["purchase_year"] = orders["order_purchase_timestamp"].dt.year

In [20]:
orders["purchase_month"] = orders["order_purchase_timestamp"].dt.month_name()

In [21]:
orders["purchase_day"] = orders["order_purchase_timestamp"].dt.day

In [22]:
orders["purchase_weekday"] = orders["order_purchase_timestamp"].dt.day_name()

In [23]:
orders[[
    "order_purchase_timestamp",
    "purchase_year",
    "purchase_month",
    "purchase_day",
    "purchase_weekday"]].head()

,order_purchase_timestamp,purchase_year,purchase_month,purchase_day,purchase_weekday
0,2017-10-02 10:56:33,2017,October,2,Monday
1,2018-07-24 20:41:37,2018,July,24,Tuesday
2,2018-08-08 08:38:49,2018,August,8,Wednesday
3,2017-11-18 19:28:06,2017,November,18,Saturday
4,2018-02-13 21:18:39,2018,February,13,Tuesday


In [24]:
orders["is_weekend"] = orders["purchase_weekday"].isin(["Saturday","Sunday"])

In [25]:
orders["delivery_delay_days"] = (
    orders["order_delivered_customer_date"] -
    orders["order_estimated_delivery_date"]
).dt.days

In [26]:
orders["is_delayed"] = orders["delivery_delay_days"] > 0

In [27]:
orders[[
    "purchase_weekday",
    "is_weekend",
    "delivery_delay_days",
    "is_delayed"
]].head()

,purchase_weekday,is_weekend,delivery_delay_days,is_delayed
0,Monday,False,-8.0,False
1,Tuesday,False,-6.0,False
2,Wednesday,False,-18.0,False
3,Saturday,True,-13.0,False
4,Tuesday,False,-10.0,False


In [225]:
orders["is_weekend"] = orders["is_weekend"].astype(int)
orders["is_delayed"] = orders["is_delayed"].astype(int)

In [28]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,purchase_year,purchase_month,purchase_day,purchase_weekday,is_weekend,delivery_delay_days,is_delayed
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,8.0,2017,October,2,Monday,False,-8.0,False
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,13.0,2018,July,24,Tuesday,False,-6.0,False
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,9.0,2018,August,8,Wednesday,False,-18.0,False
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,13.0,2017,November,18,Saturday,True,-13.0,False
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,2.0,2018,February,13,Tuesday,False,-10.0,False


## Feature Engineering - Orders Table

Created the following business features:

- delivery_days
- purchase_year
- purchase_month
- purchase_day
- purchase_weekday
- is_weekend
- delivery_delay_days
- is_delayed

### Business Value

These features support:
- Monthly and yearly trend analysis
- Weekday vs weekend sales analysis
- Delivery performance tracking
- SLA (Service Level Agreement) monitoring
- On-time delivery reporting

In [227]:
#Count delayed orders
orders["is_delayed"].value_counts()

is_delayed
0    92906
1     6535
Name: count, dtype: int64

In [229]:
orders["is_weekend"].value_counts()

is_weekend
0    76594
1    22847
Name: count, dtype: int64

In [32]:
month_order = [
    "January","February","March","April",
    "May","June","July","August",
    "September","October","November","December"
]

orders["purchase_month"] = pd.Categorical(
    orders["purchase_month"],
    categories=month_order,
    ordered=True
)

In [33]:
orders["purchase_month"].value_counts().sort_index()

purchase_month
January       8069
February      8508
March         9893
April         9343
May          10573
June          9412
July         10318
August       10843
September     4305
October       4959
November      7544
December      5674
Name: count, dtype: int64

In [233]:
orders.to_csv(
    r"D:\Retail 360\Dataset\Silver\orders_clean.csv",
    index=False
)

# customers table

In [36]:
customers = pd.read_csv(r"D:\Retail 360\Dataset\Raw\customers.csv")

In [37]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [38]:
customers.shape

(99441, 5)

In [39]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 11.0 MB


In [40]:
customers.dtypes

customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

In [41]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [42]:
customers.duplicated().sum()

0

In [43]:
customers.nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

In [44]:
customers["customer_unique_id"].duplicated().sum()

3345

In [45]:
customers["customer_unique_id"].value_counts().head(10)

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
dc813062e0fc23409cd255f7f53c7074     6
63cfc61cee11cbe306bff5857d00bfe4     6
f0e310a6839dce9de1638e0fe5ab282a     6
de34b16117594161a6a89c50b289d35a     6
Name: count, dtype: int64

In [46]:
customers["customer_city"] = customers["customer_city"].str.title()

In [47]:
customers.to_csv(
    r"D:\Retail 360\Dataset\Silver\customers_clean.csv",
    index=False
)

## Customers Table - Cleaning Summary

### Dataset Overview
- Rows: 99,441
- Columns: 5

### Data Quality Checks
- Missing Values: None
- Duplicate Rows: None

### Cleaning Performed
- Standardized customer_city using title case.

### Observations
- customer_id is unique for every order.
- customer_unique_id identifies the actual customer.
- Some customers have multiple customer_id values, indicating repeat purchases.

# order_items table

In [50]:
order_items = pd.read_csv(r"D:\Retail 360\Dataset\Raw\order_items.csv")

In [51]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [52]:
order_items.shape

(112650, 7)

In [53]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 18.4 MB


In [54]:
order_items.dtypes

order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

In [55]:
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [56]:
order_items.duplicated().sum()

0

In [57]:
order_items.nunique()

order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [58]:
order_items["shipping_limit_date"] = pd.to_datetime(order_items["shipping_limit_date"])

In [59]:
order_items.dtypes

order_id                          str
order_item_id                   int64
product_id                        str
seller_id                         str
shipping_limit_date    datetime64[us]
price                         float64
freight_value                 float64
dtype: object

In [60]:
(order_items["price"] < 0).sum()

0

In [61]:
(order_items["price"] == 0).sum()

0

In [62]:
(order_items["freight_value"] < 0).sum()

0

In [63]:
(order_items["freight_value"] == 0).sum()

383

In [64]:
order_items["total_amount"] = order_items["price"] + order_items["freight_value"]

In [65]:
order_items[[ "price", "freight_value", "total_amount"]].head()

,price,freight_value,total_amount
0,58.90,13.29,72.19
1,239.90,19.93,259.83
2,199.00,17.87,216.87
3,12.99,12.79,25.78
4,199.90,18.14,218.04


In [66]:
order_items["freight_percentage"] = (order_items["freight_value"] / order_items["price"]) * 100

In [67]:
order_items["is_expensive"] = order_items["price"] > 500

In [235]:
order_items["is_expensive"] = order_items["is_expensive"].astype(int)

## Order Items - Cleaning Summary

### Data Quality
- Rows: 112,650
- Columns: 7
- Missing Values: None
- Duplicate Rows: None

### Cleaning Performed
- Converted shipping_limit_date to datetime.

### Feature Engineering
- total_amount
- freight_percentage
- is_expensive

### Business Value
- Enables revenue analysis.
- Enables shipping cost analysis.
- Supports premium product segmentation.

### Business Validation
- No negative prices found.
- No negative freight charges found.
- 383 orders have zero freight charges, which are retained as valid business cases (e.g., free shipping or shipping included in product price).e).oduct segmentation.

In [237]:
order_items.to_csv(
    r"D:\Retail 360\Dataset\Silver\order_items_clean.csv",
    index=False
)

# products table

In [71]:
products = pd.read_csv(r"D:\Retail 360\Dataset\Raw\products.csv")

In [72]:
products.head()

,product_id,product_category_name,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [73]:
products.shape

(32951, 9)

In [74]:
products.dtypes

product_id                        str
product_category_name             str
product_name_length           float64
product_description_length    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [75]:
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_length         32341 non-null  float64
 3   product_description_length  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 3.7 MB


In [76]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_length           610
product_description_length    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [77]:
products.duplicated().sum()

0

In [78]:
products.nunique()

product_id                    32951
product_category_name            73
product_name_length              66
product_description_length     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

In [79]:
products["product_weight_g"] = products["product_weight_g"].fillna(products["product_weight_g"].median())

In [80]:
products["product_length_cm"] = products["product_length_cm"].fillna(products["product_length_cm"].median())

In [81]:
products["product_height_cm"] = products["product_height_cm"].fillna(products["product_height_cm"].median())

In [82]:
products["product_width_cm"] = products["product_width_cm"].fillna(products["product_width_cm"].median())

In [83]:
products["product_category_name"] = products["product_category_name"].fillna("Unknown")

In [84]:
products["product_volume_cm3"] = products["product_length_cm"] * products["product_height_cm"] * products["product_width_cm"]

In [85]:
products["is_heavy"] = products["product_weight_g"] > 5000

In [86]:
products["is_large"] = products["product_volume_cm3"] > 100000

In [221]:
products["is_heavy"] = products["is_heavy"].astype(int)
products["is_large"] = products["is_large"].astype(int)

## Products Table - Cleaning Summary

### Data Quality
- Rows: 32,951
- Columns: 9
- Duplicate Rows: 0

### Missing Values
- Filled missing product categories with "Unknown".
- Filled missing product dimensions and weight using the median.
- Retained missing values in product name length, description length, and photo count because the true values are unknown.

### Feature Engineering
- product_volume_cm3
- is_heavy
- is_large

### Business Value
- Enables logistics analysis.
- Supports shipping cost analysis.
- Helps identify large and heavy products.

In [88]:
products.isnull().sum()

product_id                      0
product_category_name           0
product_name_length           610
product_description_length    610
product_photos_qty            610
product_weight_g                0
product_length_cm               0
product_height_cm               0
product_width_cm                0
product_volume_cm3              0
is_heavy                        0
is_large                        0
dtype: int64

In [89]:
#data quality checks

In [90]:
products["product_weight_g"].describe()

count    32951.000000
mean      2276.376802
std       4281.926387
min          0.000000
25%        300.000000
50%        700.000000
75%       1900.000000
max      40425.000000
Name: product_weight_g, dtype: float64

In [91]:
products["product_length_cm"].describe()

count    32951.000000
mean        30.814725
std         16.914005
min          7.000000
25%         18.000000
50%         25.000000
75%         38.000000
max        105.000000
Name: product_length_cm, dtype: float64

In [92]:
products["product_height_cm"].describe()

count    32951.000000
mean        16.937422
std         13.637175
min          2.000000
25%          8.000000
50%         13.000000
75%         21.000000
max        105.000000
Name: product_height_cm, dtype: float64

In [93]:
products["product_width_cm"].describe()

count    32951.000000
mean        23.196534
std         12.078707
min          6.000000
25%         15.000000
50%         20.000000
75%         30.000000
max        118.000000
Name: product_width_cm, dtype: float64

In [223]:
products.to_csv(
    r"D:\Retail 360\Dataset\Silver\products_clean.csv",
    index=False
)

# payment table

In [96]:
payment = pd.read_csv(r"D:\Retail 360\Dataset\Raw\order_payments.csv")

In [97]:
payment.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [98]:
payment.shape

(103886, 5)

In [99]:
payment.dtypes

order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

In [100]:
payment.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 8.1 MB


In [101]:
payment.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [102]:
payment.duplicated().sum()

0

In [103]:
payment.nunique()

order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

In [104]:
payment["installments_cateogry"] = payment["payment_installments"].apply(
    lambda x: "Single Payment" if x == 1 else "EMI"
    )

In [105]:
payment["high_value_payments"] = payment["payment_value"] > 1000

In [106]:
payment["is_emi"] = payment["payment_installments"] > 1

In [252]:
payment["is_emi"] = payment["is_emi"].astype(int)
payment["high_value_payments"] = payment["high_value_payments"].astype(int)

# Payments Table - Cleaning Summary

## Dataset Overview

- Rows: 103,886
- Colum

---

## Data Quality Checks

- Missing Values: None
- Duplicate Rows: None
- Negative Payment Values: None
- Invalid Instnts: None

---

## Cleaning Performed

- No major cleaning was required as the dataset had no missing values or duplicate records.
- Validated payment values to ensure there were no negative amounts.
- Validated installment values to ensure they wered numeric values.

---

## Feature Engineering

### 1. installment_category

Categorized customers based on the number of installments:

- Single Payment (1 installment)
- EMI (More than 1 installment)

### 2. is_emi

Created a Boolean column indicating whether the payment was made using installments.

- True → Installment payment
- False → Single payment

### 3. high_value_payment

Created a Boolean column to identify high-value transactions.

- True → Paymenue > 1000
- False → Payment Value ≤ 1000

---

## Business Value

The engineered features enable:

- Payment method analysis
- EMI adoption analysis
- High-value customer identification
- Revenue analysis by payment type
- Insted dataset was saved to the Silver layer as:

`payments_clean.csv`

In [254]:
payment.to_csv(
    r"D:\Retail 360\Dataset\Silver\payment_clean.csv",
    index=False
)

In [109]:
orders[["purchase_year", "purchase_month"]].head(10)

,purchase_year,purchase_month
0,2017,October
1,2018,July
2,2018,August
3,2017,November
4,2018,February
5,2017,July
6,2017,April
7,2017,May
8,2017,January
9,2017,July


In [110]:
orders["order_purchase_timestamp"].head()

0   2017-10-02 10:56:33
1   2018-07-24 20:41:37
2   2018-08-08 08:38:49
3   2017-11-18 19:28:06
4   2018-02-13 21:18:39
Name: order_purchase_timestamp, dtype: datetime64[us]

In [111]:
orders["order_purchase_timestamp"].dtype

dtype('<M8[us]')

In [112]:
orders["order_purchase_timestamp"].dt.month_name().head(10)

0     October
1        July
2      August
3    November
4    February
5        July
6       April
7         May
8     January
9        July
Name: order_purchase_timestamp, dtype: str

In [113]:
orders["order_purchase_timestamp"].head(10)

0   2017-10-02 10:56:33
1   2018-07-24 20:41:37
2   2018-08-08 08:38:49
3   2017-11-18 19:28:06
4   2018-02-13 21:18:39
5   2017-07-09 21:57:05
6   2017-04-11 12:22:08
7   2017-05-16 13:10:30
8   2017-01-23 18:29:09
9   2017-07-29 11:55:02
Name: order_purchase_timestamp, dtype: datetime64[us]

In [114]:
orders["order_purchase_timestamp"].dt.month.head(10)

0    10
1     7
2     8
3    11
4     2
5     7
6     4
7     5
8     1
9     7
Name: order_purchase_timestamp, dtype: int32

In [115]:
orders["purchase_month"].head(10)

0     October
1        July
2      August
3    November
4    February
5        July
6       April
7         May
8     January
9        July
Name: purchase_month, dtype: category
Categories (12, str): ['January' < 'February' < 'March' < 'April' ... 'September' < 'October' < 'November' < 'December']

In [250]:
print(payment.columns.tolist())

['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value', 'installments_cateogry', 'high_value_payments', 'is_emi']
